Added E2E tests for sidebar navigation during Ember-to-React migration

rob-ghost opened the pull request:
ref https://linear.app/ghost/issue/BER-3038

## What does this do?

Adds E2E tests for the admin sidebar and updates markup to support testing with semantic selectors:

- Added E2E tests covering main navigation, posts submenu, custom views, and footer interactions
- Updated markup in both Ember and React sidebars with ARIA attributes (`aria-current`, `aria-expanded`, `aria-label`, `role="menuitem"`) to enable semantic locators
- Removed Ember sidebar when running in React shell

These have been segmented into 3 separate commits for easier reviewing if you want to review each change in isolation.

## Why?

We need stable tests during the Ember-to-React migration. Using semantic selectors based on ARIA attributes allows tests to work with both implementations regardless of markup changes.

## ❌ Known Failing Tests

The React tests for the active nav items are currently failing as the logic is implemented in https://github.com/TryGhost/Ghost/pull/25518. The React e2e tests are still currently optional and do not block merge on failure.

## ⏩ Skipped Dark Mode tests

Dark mode test was skipped because of differences in the markup between Ember and React. We need to decide if we:
1. Keep it skipped until we settle on a markup for dark mode (user menu or sidebar toggle)
2. Conditionally skip it depending on the shell
3. Have two tests, one for Ember and React, skipped conditionally depending on the shell
4. Custom code in the Sidebar page object to deal with the differences in markup

rob-ghost committed:
Improved sidebar accessibility with ARIA attributes

ref https://linear.app/ghost/issue/BER-3038

Added proper ARIA attributes to both React and Ember sidebar components for better screen reader support and assistive technology compatibility. This ensures proper semantic structure for menus and navigation during the Ember to React migration.

rob-ghost committed:
Removed Ember sidebar when hosted in React shell

ref https://linear.app/ghost/issue/BER-3038

When running Ember admin inside the React shell, the Ember sidebar is now properly hidden to prevent duplicate sidebars in the DOM. Also moved keyboard shortcut handling from component to application controller to ensure shortcuts work regardless of which sidebar is active.

rob-ghost committed:
Added E2E tests for sidebar navigation

ref https://linear.app/ghost/issue/BER-3038

Comprehensive E2E tests covering sidebar navigation, custom post views, and footer interactions. Tests use ARIA-based locators to work seamlessly with both Ember and React sidebar implementations during the migration.

jonatansberg commented on apps/shade/src/components/ui/sidebar.tsx:
```diff
@@ -219,6 +219,7 @@ const Sidebar = React.forwardRef<
                 data-side={side}
                 data-state={state}
                 data-variant={variant}
+                role="navigation"
```
I think we can go ahead and add this everywhere!

jonatansberg commented on ghost/admin/app/components/gh-nav-menu/main.hbs:
```diff
@@ -61,17 +61,19 @@
                     {{#if this.session.user.isAuthorOrContributor}}
                     {{#if this.customViews.forPosts}}
-                    <ul class="gh-nav-view-list">
+                    <ul class="gh-nav-view-list" id="posts-submenu">
```
Remove this if we can have a better locator

jonatansberg commented on e2e/helpers/pages/admin/sidebar/sidebar-page.ts:
```diff
@@ -0,0 +1,128 @@
+
+    getVisibleItemsForRole(role: UserRole): NavItem[] {
+        return SidebarPage.NAV_ITEMS.filter(item => item.roles.includes(role));
+    }
```
This seems to be unused

jonatansberg commented on e2e/tests/admin/sidebar/navigation.test.ts:
```diff
@@ -0,0 +1,161 @@
+
+                await sidebar.clickNavItem(name);
+
+                await expect(page).toHaveURL(new RegExp(expectedPath.replace(/[#]/g, '\\$&')));
```
It would be nice if we can clean up this regex and not do a bunch of string manipulation. Tweaking the expected paths above would allow us to do so.

jonatansberg reviewed (commented):
Here are the comments from or mob review session

rob-ghost committed:
Added E2E tests for custom post views

ref https://linear.app/ghost/issue/BER-3038

Added test coverage for custom post views including creation, navigation, editing, deletion, filter modifications, and persistence. Tests use semantic locators and wait for page state to settle before interacting with filters to avoid race conditions.

ibalosh commented on e2e/helpers/pages/admin/posts/custom-view-modal.ts:
```diff
@@ -0,0 +1,48 @@
+        this.page = page;
+        this.modal = page.getByRole('dialog');
+        this.nameInput = page.getByLabel('View name');
+        this.nameError = page.locator('[data-test-error="custom-view-name"]');
```
I assume later this one will end up being `page.getByTestId("error-customer-view")` or something like that

ibalosh commented on e2e/helpers/pages/admin/posts/posts-page.ts:
```diff
@@ -13,13 +24,75 @@ export class PostsPage extends AdminPage {
     }
 
+    async waitForPostsToLoad(): Promise<void> {
+        await Promise.race([
```
`nit:` maybe we could describe the intention of this wait function with a comment, so it's clear why we handle the race condition and silence rejections

ibalosh replied:
`nit` and maybe make it even clearer with named locators:

```
await Promise.race([
    firstPost.waitFor({state: 'visible'}).catch(() => {}),
    emptyState.waitFor({state: 'visible'}).catch(() => {})
]);
```

ibalosh commented on e2e/helpers/pages/admin/sidebar/sidebar-page.ts:
```diff
@@ -0,0 +1,73 @@
+        this.sidebar = page.getByRole('navigation');
+        this.postsToggle = this.sidebar.getByRole('button', {name: /toggle post views/i});
+        this.userDropdownTrigger = page.locator('[data-test-nav="arrow-down"]');
+        this.nightShiftToggle = page.getByRole('button', {name: /dark mode/i}).or(page.getByRole('menuitem', {name: /dark mode/i}).getByRole('switch'));
```
`nit` since we use `or` for locating here, will this always work, will it find correct locators (I assume so just a double check here)

ibalosh reviewed (approved):
Couple of nitpicks and questions around locator choices (data-testids mostly). I don't have strong feelings against it though

rob-ghost commented on e2e/tests/admin/sidebar/navigation.test.ts:
```diff
@@ -0,0 +1,152 @@
+            test(`clicking ${name} - navigates and shows active state`, async ({page}) => {
+                const sidebar = new SidebarPage(page);
+
+                await sidebar.goto('/ghost');
```
Valid suggestion, but I'll defer it for now.

rob-ghost commented on e2e/helpers/pages/admin/posts/posts-page.ts:
```diff
@@ -13,13 +24,75 @@ export class PostsPage extends AdminPage {
+    }
+
     async refreshData() {
         await this.page.reload();
```
The change to align this was out-of-scope of this PR, but a good cleanup 🪣 we can do later
